File: split/scanmath.pas
procedure scanmath(p: halfword);
label
  20, 21, 10;
var
  c: integer;
begin
  20:
      {404:}
repeat
      getxtoken;
    until (curcmd <> 10) and (curcmd <> 0){:404}
;
  21:
    case curcmd of
      11, 12, 68:
      begin
        c := eqtb[5007 + curchr].hh.rh - 0;
        if c = 32768 then
        begin{1152:}

          begin
            curcs := curchr + 1;
            curcmd := eqtb[curcs].hh.b0;
            curchr := eqtb[curcs].hh.rh;
            xtoken;
            backinput;
          end{:1152}
;
          goto 20;
        end;
      end;
      16:
      begin
        scancharnum;
        curchr := curval;
        curcmd := 68;
        goto 21;
      end;
      17:
      begin
        scanfifteenbitint;
        c := curval;
      end;
      69: c := curchr;
      15:
      begin
        scantwentysevenbitint;
        c := curval div 4096;
      end;
      others:{1153:}

      begin
        backinput;
        scanleftbrace;
        savestack[saveptr + 0].int := p;
        saveptr := saveptr + 1;
        pushmath(9);
        goto 10;
      end{:1153}

    end;
  mem[p].hh.rh := 1;
  mem[p].hh.b1 := c mod 256 + 0;
  if (c >= 28672) and ((eqtb[5307].int >= 0) and (eqtb[5307].int < 16)) then
    mem[p].hh.b0 := eqtb[5307].int
  else
    mem[p].hh.b0 := (c div 256) mod 16;
  10: ;
end;

File: cpp/scanmath.cpp
#include "scanmath.h"
#include "getxtoken.h"
#include "xtoken.h"
#include "backinput.h"
#include "scancharnum.h"
#include "scanfifteenbitint.h"
#include "scantwentysevenbitint.h"
#include "backinput.h"
#include "scanleftbrace.h"
#include "pushmath.h"

void scanmath(halfword p)
{
	do
		getxtoken();
	while (curcmd == spacer || curcmd == escape);
	bool label21;
	do
	{
		label21 = false;
		int c;
		switch (curcmd)
		{
			case letter:
			case other_char:
			case char_given:
				c = math_code(curchr);
				if (c == 0x80'00)
				{
					curcs = curchr+1;
					curcmd = eq_type(curcs);
					curchr = equiv(curcs);
					xtoken();
					backinput();
					do
						getxtoken();
					while (curcmd == spacer || curcmd == escape);
					label21 = true;
					continue;
				}
				break;
			case char_num:
				scancharnum();
				curchr = curval;
				curcmd = char_given;
				label21 = true;
				continue;
			case math_char_num:
				scanfifteenbitint();
				c = curval;
				break;
			case math_given: 
				c = curchr;
				break;
			case delim_num:
				scantwentysevenbitint();
				c = curval/0x10'00;
				break;
			default:
				backinput();
				scanleftbrace();
				savestack[saveptr++].int_ = p;
				pushmath(9);
				return;
		}
	} while (label21);
	link(p) = 1;
	subtype(p) = c%0x1'00;
	if (c >= 0x7000 && int_par(cur_fam_code) >= 0 && int_par(cur_fam_code) < 0x10)
		type(p) = int_par(cur_fam_code);
	else
		type(p) = (c/0x1'00)%0x10;
}
